cirq/algorithms/qaoa/QAOA.py

```diff
@@ -0,0 +1,353 @@
+        """
+        self.objective_function = objective_function
+
+    def set_mixing_operator(self, mixing_operator=None):
+        """
+        sets new mixing hermitian operator (B in QAOA paper) that will be
+        used to construct the unitary U_B(\beta) = Exp(-i*beta*B)
+
+        :param mixing_operator: Dict[PauliString: Coefficient]
+            operator B in QAOA paper, used to create unitary by exponentiation
+            default set to 'X' which sets it to pauli H on every qubit
+        """
+
+        # initialize operator dict
+        self.mixing_operator = {}
+
+        # sets it to X on every qubit
+        if mixing_operator == None:
+            # X on every qubit
+            for i in range(self.n_qubits):
+                self.mixing_operator[PauliString(
+                    qubit_pauli_map={self.qubits[i]:Pauli.X})] = 1
+
+        # sets it to argument given in function
+        else:
+            self.mixing_operator = mixing_operator
+
+    def regular_tree_max_cut_graph(self, n_connections):
+        """
+        sets up the cost function for a regular tree graph (no loops) where
+        each vertex has n_connections edges connected to it.
+
+        :param n_connections: integer determining the number of connections
+            for each vertex in the regular tree graph
+        """
+        z = Pauli.Z
+
+        # Ignores constant term since it just adds a phase
+        self.objective_function = dict()
+
+        # connect 0 to 1:
+        self.objective_function[PauliString(qubit_pauli_map={self.qubits[0]:z,
+                                            self.qubits[1]:z})] = -1/2
+
+        self.objective_function[()] = 1 / 2
+
+        # connections
+        # odd: (n_connections-1)*i + (4-n_connections + 2*j)
+        # even: (n_connections-1)*i + 2 + 2*j
+        for i in range(self.n_qubits):
+            for j in range(n_connections-1):
+
+                if i % 2 == 1:
+                    odd_connection = (n_connections-1)*i + (4 -
+                                                            n_connections + 2*j)
+                    if odd_connection < self.n_qubits:
+                        self.objective_function[PauliString(
+                            qubit_pauli_map={self.qubits[i]:z,
+                                            self.qubits[odd_connection]:z})] \
+                            = -1/2
+
+                        self.objective_function[()] += 1/2
+
+                # even qubits
+                if i % 2 == 0:
+                    even_connection = (n_connections-1)*i + 2 + 2*j
+                    if even_connection < self.n_qubits:
+                        self.objective_function[PauliString(
+                            qubit_pauli_map={self.qubits[i]:z,
+                                             self.qubits[even_connection]:z})] \
+                            = -1/2
+
+                        self.objective_function[()] += 1 / 2
```

Comment:
Both branches of the `if` end in the same statement. Put it afterwards unconditionally.

Actually, it looks like this will just be `len(self.qubits) * (n_connections - 1) / 2` total so just set the total outside the loop if that's right.

Reply:
The type of graph I'm considering here has no loops, so it should just be (len(self.qubits) - 1)/2. I removed it from the loop